casr/src/util.rs

```diff
@@ -293,3 +284,532 @@ pub fn get_output(command: &mut Command, timeout: u64, error_on_timeout: bool) -
+fn initialize_dirs(matches: &clap::ArgMatches) -> Result<(&PathBuf, PathBuf, PathBuf)> {
+    // Get output dir
+    let output_dir = matches.get_one::<PathBuf>("output").unwrap();
+    if !output_dir.exists() {
+        fs::create_dir_all(output_dir).with_context(|| {
+            format!("Couldn't create output directory {}", output_dir.display())
+        })?;
+    } else if output_dir.read_dir()?.next().is_some() {
+        bail!("Output directory is not empty.");
+    }
+    // Get oom dir
+    let oom_dir = output_dir.join("oom");
+    if fs::create_dir_all(&oom_dir).is_err() {
+        bail!("Failed to create dir {}", &oom_dir.to_str().unwrap());
+    }
+    // Get timeout dir
+    let timeout_dir = output_dir.join("timeout");
+    if fs::create_dir_all(&timeout_dir).is_err() {
+        bail!("Failed to create dir {}", &timeout_dir.to_str().unwrap());
+    }
+
+    Ok((output_dir, oom_dir, timeout_dir))
+}
+
+#[derive(Debug, Clone, Default)]
+/// Information about crash to reproduce it.
+pub struct CrashInfo {
+    /// Path to crash input.
+    pub path: PathBuf,
+    /// Target command line args.
+    pub target_args: Vec<String>,
+    /// Input file argument index starting from argv\[1\], None for stdin.
+    pub at_index: Option<usize>,
+    /// ASAN.
+    pub is_asan: bool,
+}
+
+impl<'a> CrashInfo {
+    /// Generate Casr report for crash.
+    ///
+    /// # Arguments
+    ///
+    /// * `tool` - path to tool that generates reports
+    ///
+    /// * `output_dir` - save report to specified directory or use the same directory as crash
+    ///
+    /// * `timeout` - target program timeout (in seconds)
+    ///
+    /// *  `envs` - environment variables for target
+    pub fn run_casr<T: Into<Option<&'a Path>>>(
+        &self,
+        tool: &PathBuf,
+        output_dir: T,
+        timeout: u64,
+        envs: &HashMap<String, String>,
+    ) -> Result<()> {
+        let mut args: Vec<String> = vec!["-o".to_string()];
+        let (report_path, output_dir) = if let Some(out) = output_dir.into() {
+            (out.join(self.path.file_name().unwrap()), out)
+        } else {
+            (self.path.clone(), self.path.parent().unwrap())
+        };
+        if self.is_asan {
+            args.push(format!("{}.casrep", report_path.display()));
+        } else {
+            args.push(format!("{}.gdb.casrep", report_path.display()));
+        }
+
+        args.push("--".to_string());
+        args.extend_from_slice(&self.target_args);
+        if let Some(at_index) = self.at_index {
+            let input = args[at_index + 3].replace("@@", self.path.to_str().unwrap());
+            args[at_index + 3] = input;
+        }
+        let tool_name = tool.file_name().unwrap().to_str().unwrap();
+        if tool_name.ends_with("casr-python") {
+            args.insert(3, "python3".to_string());
+        }
+        if self.at_index.is_none() {
+            args.insert(0, self.path.to_str().unwrap().to_string());
```

Comment:
Is it possible rewrite it without vec shifting?

Reply:
Seems impossible. In this case we will have problems with insertion of casr-python.

Reply:
reworked